### PR TITLE
[motherless] Add support for groups

### PR DESCRIPTION
**Pull request by mweinelt**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add support for group URLs to be downloaded as playlists.

### CODE REVIEW

File: youtube_dl/extractor/motherless.py [hunk 1]
```diff
@@ -114,3 +116,72 @@ def _real_extract(self, url):
             'age_limit': age_limit,
             'url': video_url,
         }
+
+
+class MotherlessGroupIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?motherless\.com/gv?/(?P<id>[a-z0-9_]+)/?$'
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: `/?$` senseless. Does not match http://motherless.com/gv/sex_must_be_funny?foo=bar.

File: youtube_dl/extractor/motherless.py [hunk 2]
```diff
@@ -114,3 +116,72 @@ def _real_extract(self, url):
...
+                r'href="/([^"]+)"[^>]+>\s+<img[^>]+alt="[^-]+-\s([^"]+)"',
+                webpage))
+        ]
+
+    def _real_extract(self, url):
+        group_id = self._match_id(url)
+        webpage = self._download_webpage(
+            'http://motherless.com/gv/%s' % group_id, group_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+)', webpage, 'title', fatal=True).strip()
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Playlist title should not be fatal.

File: youtube_dl/extractor/motherless.py [hunk 3]
```diff
@@ -114,3 +116,72 @@ def _real_extract(self, url):
...
+        description = self._search_regex(
+            r'<meta name="description" content="([^"]+)">',
+            webpage, 'description', fatal=False)
+        page_count = self._int(self._search_regex(
+            r'(\d+)</(a|span)><(a|span)[^>]+>\s*NEXT',
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Don't capture groups you don't use.

File: youtube_dl/extractor/motherless.py [hunk 4]
```diff
@@ -114,3 +116,72 @@ def _real_extract(self, url):
...
+            webpage, 'page_count'), 'page_count', fatal=True)
+        page_size = 80
+
+        def _get_page(idx):
+            page_url = 'http://motherless.com/gv/%s?page=%d' % (
+                group_id, idx + 1
+            )
+            webpage = self._download_webpage(
+                page_url, group_id,
+                note="Downloding page %d/%d" % (idx + 1, page_count)
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Single quotes.

File: youtube_dl/extractor/motherless.py [hunk 5]
```diff
@@ -114,3 +116,72 @@ def _real_extract(self, url):
...
+            'description': 'Sex can be funny. Wide smiles,laugh, games, fun of '
+                           'any kind!'
+        },
+        'playlist_mincount': 9,
+    }, {
+        'url': 'http://motherless.com/g/movie_scenes',
+        'info_dict': {
+            'id': 'movie_scenes',
+            'title': 'Movie Scenes',
+            'description': 'Hot and sexy scenes from &quot;regular&quot; '
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Not unescaped.

File: youtube_dl/extractor/motherless.py (hunk 1 above)
Comment (mweinelt) [reply youtube_dl/extractor/motherless.py]: It may not match `http://motherless.com/g/cosplay/633979F` and I'm a bit at a loss on how to accomplish that.

**Comment by mweinelt** [comment]
issues addressed.

File: youtube_dl/extractor/motherless.py [hunk 6]
```diff
@@ -114,3 +117,72 @@ def _real_extract(self, url):
...
+                r'href="/([^"]+)"[^>]+>\s+<img[^>]+alt="[^-]+-\s([^"]+)"',
+                webpage))
+        ]
+
+    def _real_extract(self, url):
+        group_id = self._match_id(url)
+        webpage = self._download_webpage(
+            'http://motherless.com/gv/%s' % group_id, group_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+)', webpage, 'title', fatal=False).strip()
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Breaks on `None` title.

File: youtube_dl/extractor/motherless.py [hunk 7]
```diff
@@ -114,3 +117,72 @@ def _real_extract(self, url):
             'age_limit': age_limit,
             'url': video_url,
         }
+
+
+class MotherlessGroupIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?motherless\.com/gv?/(?P<id>[a-z0-9_]+)(?:$|/[^A-F0-9]|/?\?)'
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Overload `suitable` instead.

File: youtube_dl/extractor/motherless.py [hunk 8]
```diff
@@ -114,3 +117,72 @@ def _real_extract(self, url):
...
+                MotherlessIE.ie_key(), video_title=title)
+            for video_url, title in orderedSet(re.findall(
+                r'href="/([^"]+)"[^>]+>\s+<img[^>]+alt="[^-]+-\s([^"]+)"',
+                webpage))
+        ]
+
+    def _real_extract(self, url):
+        group_id = self._match_id(url)
+        webpage = self._download_webpage(
+            'http://motherless.com/gv/%s' % group_id, group_id)
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Use the original scheme and host.

File: youtube_dl/extractor/motherless.py [hunk 9]
```diff
@@ -114,3 +117,72 @@ def _real_extract(self, url):
...
+                           'Beautiful actresses fully nude... A looot of '
+                           'skin! :)Enjoy!',
+        },
+        'playlist_mincount': 662,
+    }]
+
+    def _extract_entries(self, webpage):
+        return [
+            self.url_result(
+                'http://www.motherless.com/%s' % video_url,
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Use the original scheme and host.

File: youtube_dl/extractor/motherless.py [hunk 10]
```diff
@@ -114,3 +117,72 @@ def _real_extract(self, url):
...
+        description = unescapeHTML(self._search_regex(
+            r'<meta name="description" content="([^"]+)">',
+            webpage, 'description', fatal=False))
+        page_count = self._int(self._search_regex(
+            r'(\d+)</(?:a|span)><(?:a|span)[^>]+>\s*NEXT',
+            webpage, 'page_count'), 'page_count', fatal=True)
+        page_size = 80
+
+        def _get_page(idx):
+            page_url = 'http://motherless.com/gv/%s?page=%d' % (
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Query to `query`,

**Comment by mweinelt** [comment]
Anything else?

File: youtube_dl/extractor/motherless.py [hunk 11]
```diff
@@ -114,3 +118,81 @@ def _real_extract(self, url):
...
+        ]
+
+    def _real_extract(self, url):
+        parsed_url = compat_urlparse.urlparse(url)
+        http_scheme = parsed_url.scheme
+        http_host = parsed_url.netloc
+
+        group_id = self._match_id(url)
+        webpage = self._download_webpage(
+            '%s://%s/gv/%s' % (http_scheme, http_host, group_id), group_id)
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: urljoin.

File: youtube_dl/extractor/motherless.py [hunk 12]
```diff
@@ -114,3 +118,81 @@ def _real_extract(self, url):
...
+        description = unescapeHTML(self._search_regex(
+            r'<meta name="description" content="([^"]+)">',
+            webpage, 'description', fatal=False))
+        page_count = self._int(self._search_regex(
+            r'(\d+)</(?:a|span)><(?:a|span)[^>]+>\s*NEXT',
+            webpage, 'page_count'), 'page_count', fatal=True)
+        page_size = 80
+
+        def _get_page(idx):
+            page_url = '%s://%s/gv/%s' % (
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: urljoin.

File: youtube_dl/extractor/motherless.py [hunk 13]
```diff
@@ -114,3 +118,81 @@ def _real_extract(self, url):
...
+
+    @classmethod
+    def suitable(cls, url):
+        return (False if MotherlessIE.suitable(url)
+                else super(MotherlessGroupIE, cls).suitable(url))
+
+    def _extract_entries(self, webpage, http_scheme, http_host):
+        return [
+            self.url_result(
+                '%s://%s/%s' % (http_scheme, http_host, video_url),
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: urljoin.

File: youtube_dl/extractor/motherless.py [hunk 14]
```diff
@@ -114,3 +118,81 @@ def _real_extract(self, url):
...
+        webpage = self._download_webpage(
+            '%s://%s/gv/%s' % (http_scheme, http_host, group_id), group_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+\w)\s+-', webpage, 'title', fatal=False)
+        description = unescapeHTML(self._search_regex(
+            r'<meta name="description" content="([^"]+)">',
+            webpage, 'description', fatal=False))
+        page_count = self._int(self._search_regex(
+            r'(\d+)</(?:a|span)><(?:a|span)[^>]+>\s*NEXT',
+            webpage, 'page_count'), 'page_count', fatal=True)
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: fatal=True is default.

File: youtube_dl/extractor/motherless.py [hunk 15]
```diff
@@ -114,3 +118,81 @@ def _real_extract(self, url):
...
+        page_size = 80
+
+        def _get_page(idx):
+            page_url = '%s://%s/gv/%s' % (
+                http_scheme, http_host, group_id
+            )
+            webpage = self._download_webpage(
+                page_url, group_id, query={'page': idx + 1},
+                note='Downloding page %d/%d' % (idx + 1, page_count)
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Spelling.

File: youtube_dl/extractor/motherless.py [hunk 16]
```diff
@@ -114,3 +118,81 @@ def _real_extract(self, url):
...
+            '%s://%s/gv/%s' % (http_scheme, http_host, group_id), group_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+\w)\s+-', webpage, 'title', fatal=False)
+        description = unescapeHTML(self._search_regex(
+            r'<meta name="description" content="([^"]+)">',
+            webpage, 'description', fatal=False))
+        page_count = self._int(self._search_regex(
+            r'(\d+)</(?:a|span)><(?:a|span)[^>]+>\s*NEXT',
+            webpage, 'page_count'), 'page_count', fatal=True)
+        page_size = 80
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Uppercase.

File: youtube_dl/extractor/motherless.py [hunk 17]
```diff
@@ -114,3 +118,81 @@ def _real_extract(self, url):
...
+        http_scheme = parsed_url.scheme
+        http_host = parsed_url.netloc
+
+        group_id = self._match_id(url)
+        webpage = self._download_webpage(
+            '%s://%s/gv/%s' % (http_scheme, http_host, group_id), group_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+\w)\s+-', webpage, 'title', fatal=False)
+        description = unescapeHTML(self._search_regex(
+            r'<meta name="description" content="([^"]+)">',
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: _html_search_meta.

**Comment by mweinelt** [comment]
Addressed, what's next? 😆

File: youtube_dl/extractor/motherless.py [hunk 18]
```diff
@@ -114,3 +117,76 @@ def _real_extract(self, url):
...
+                compat_urlparse.urljoin(base_url, video_path),
+                MotherlessIE.ie_key(), video_title=title)
+            for video_path, title in orderedSet(re.findall(
+                r'href="/([^"]+)"[^>]+>\s+<img[^>]+alt="[^-]+-\s([^"]+)"',
+                webpage))
+        ]
+
+    def _real_extract(self, url):
+        parsed_url = compat_urlparse.urlparse(url)
+        base_url = '%s://%s' % (parsed_url.scheme, parsed_url.netloc)
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: What's the point of this? Use `url` as base.

File: youtube_dl/extractor/motherless.py [hunk 19]
```diff
@@ -114,3 +117,76 @@ def _real_extract(self, url):
...
+            r'<title>([\w\s]+\w)\s+-', webpage, 'title', fatal=False)
+        description = self._html_search_meta(
+            'description', webpage, fatal=False)
+        page_count = self._int(self._search_regex(
+            r'(\d+)</(?:a|span)><(?:a|span)[^>]+>\s*NEXT',
+            webpage, 'page_count'), 'page_count')
+        PAGE_SIZE = 80
+
+        def _get_page(idx):
+            page_url = compat_urlparse.urljoin(base_url, '/gv/%s' % group_id)
```
Comment (dstftw) [inline youtube_dl/extractor/motherless.py]: Code duplication 165, 176.

File: youtube_dl/extractor/motherless.py (hunk 18 above)
Comment (mweinelt) [reply youtube_dl/extractor/motherless.py]: Ah neat, I was unaware urljoin allowed overwriting of the path.

**Comment by mweinelt** [comment]
Updated and squashed.

**Comment by mweinelt** [comment]
Thanks for the thorough review! 👍

**Pull request merged**